File: sdk-ms-azure-management-demo/VodProvider.cs
using Azure.Storage.Blobs;

using Microsoft.Azure.Management.Media;
using Microsoft.Azure.Management.Media.Models;
using Microsoft.IdentityModel.Clients.ActiveDirectory;
using Microsoft.Rest;
using Microsoft.Rest.Azure.Authentication;

using VodCreatorApp.Configuration;

using Jose;

namespace VodCreatorApp
{
    public class VodProvider
    {
        private const string TransformName = "RmsTestTransform2";
        private const string StreamingEndpointName = "default";
        private const string LogoMaskLabel = "logoMask";
        private readonly AzureMediaServicesOptions _azureOptions;
        private readonly RmsOptions _rmsOptions;

        public VodProvider(RmsOptions rmsOptions, AzureMediaServicesOptions? azureOptions)
        {
            _azureOptions = azureOptions;
            _rmsOptions = rmsOptions;
        }

        public async Task CreateVod(string mediaServicesType, string inputFile)
        {
            // Create a new instance of the Media Services account
            AzureMediaServicesClient mediaService;
            string resourceGroupName;
            string accountName;
            if (mediaServicesType == "ams")
            {
                mediaService = await CreateAmsClient();
                resourceGroupName = _azureOptions!.ResourceGroupName;
                accountName = _azureOptions!.MediaServicesAccountName;
            }
            else if (mediaServicesType == "rms")
            {
                mediaService = CreateRmsClient();
                resourceGroupName = _rmsOptions.ResourceGroupName;
                accountName = _rmsOptions.MediaServicesAccountName;
            }
            else
            {
                throw new ArgumentException($"Invalid media service type: {mediaServicesType}");
            }

            try
            {
                await Run(mediaService, resourceGroupName, accountName, inputFile);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception when calling API: {0}", e.Message);
            }
        }

        public async Task Run(AzureMediaServicesClient mediaService, string resourceGroupName, string accountName, string inputFile)
        {
            // Creating a unique suffix for this test run
            string unique = Guid.NewGuid().ToString()[..13];
            string inputAssetName = $"input-{unique}";
            string overlayInputAssetName = $"input-overlay-{unique}";
            string outputAssetName = $"output-{unique}";
            string jobName = $"job-{unique}";
            string locatorName = $"locator-{unique}";
            string aes128locatorName = $"locator-{unique}-aes";
            string contentKeyPolicyName = $"aeskeypolicy-{unique}";

            JobInput jobInput;
            if (IsUrl(inputFile))
            {
                jobInput = new JobInputHttp(files: new[] { inputFile });
            }
            else
            {
                // Create input asset
                var inputAsset = await CreateAsset(mediaService, resourceGroupName, accountName, inputAssetName);
                Console.WriteLine($"Input asset created: {inputAsset.Name} (container {inputAsset.Container})");

                // Create overlay input asset
                var overlayInputAsset = await CreateAsset(mediaService, resourceGroupName, accountName, overlayInputAssetName);
                Console.WriteLine($"Input asset created: {overlayInputAsset.Name} (container {overlayInputAsset.Container})");

                // Upload video to asset
                Console.WriteLine();
                Console.WriteLine("Uploading video to input asset...");
                await UploadFileToAsset(mediaService, resourceGroupName, accountName, inputAssetName, inputFile);

                Console.WriteLine("Uploading overlay file to input asset...");
                await UploadFileToAsset(mediaService, resourceGroupName, accountName, overlayInputAssetName, "DefaultInput/tests.png");

                Console.WriteLine("Video upload completed!");
                var jobInputs = new List<JobInput> { new JobInputAsset(inputAsset.Name), new JobInputAsset(overlayInputAsset.Name, label: LogoMaskLabel) };
                jobInput = new JobInputs() { Inputs = jobInputs };
            }

            // Create output assets
            var outputAsset = await CreateAsset(mediaService, resourceGroupName, accountName, outputAssetName);
            Console.WriteLine();
            Console.WriteLine($"Output asset created: {outputAsset.Name} (container {outputAsset.Container})");

            var transform = await CreateTransformAsync(mediaService, resourceGroupName, accountName, TransformName);

            // Create job
            var job = await SubmitJobAsync(
                mediaService,
                resourceGroupName,
                accountName,
                transform.Name,
                jobName,
                jobInput,
                outputAsset.Name);
            Console.WriteLine();
            Console.WriteLine($"Job created: {job.Name}");

            // Track job progress
            job = await WaitForJobToFinishAsync(mediaService, resourceGroupName, accountName, TransformName, jobName);

            if (job.State == JobState.Error)
            {
                Console.WriteLine($"ERROR: Encoding job has failed {job.Outputs.First().Error.Message}");
                return;
            }

            Console.WriteLine($"Job finished: {job.Name}");

            // Create streaming locator
            await CreateStreamingLocatorAsync(mediaService, resourceGroupName, accountName, outputAsset.Name, locatorName);
            Console.WriteLine();
            Console.WriteLine($"Streaming locator created: {locatorName}");
            var streamingEndpoint = await mediaService.StreamingEndpoints.GetAsync(resourceGroupName, accountName, StreamingEndpointName);

            if (streamingEndpoint.ResourceState != StreamingEndpointResourceState.Running)
            {
                await mediaService.StreamingEndpoints.StartAsync(resourceGroupName, accountName, StreamingEndpointName);
            }

            // List streaming locator for asset
            var streamingLocator = await mediaService.StreamingLocators.GetAsync(resourceGroupName, accountName, locatorName);

            // List url for streaming locator
            var paths = await mediaService.StreamingLocators.ListPathsAsync(resourceGroupName, accountName, locatorName);

            var streamingUrls = new List<string>();

            Console.WriteLine();
            Console.WriteLine("The following URLs are available for adaptive streaming:");
            // All paths returned can be streamed when append it to steaming endpoint or CDN domain name
            foreach (StreamingPath path in paths.StreamingPaths)
            {
                foreach (string streamingFormatPath in path.Paths)
                {
                    var streamingUrl = $"https://{streamingEndpoint.HostName}{streamingFormatPath}";
                    Console.WriteLine($"{path.StreamingProtocol}: {streamingUrl}");
                    streamingUrls.Add(streamingUrl);
                }
            }

            Console.WriteLine();
            Console.WriteLine("The following URLs are available for downloads:");
            // All paths returned can be streamed when append it to steaming endpoint or CDN domain name
            foreach (string path in paths.DownloadPaths)
            {
                var streamingUrl = $"https://{streamingEndpoint.HostName}{path}";
                Console.WriteLine(streamingUrl);
                streamingUrls.Add(streamingUrl);
            }

            // Prepearing AES-128 encrypted HLS stream
            string issuer = "ravnur";
            string audience = "rmstest";
            // Create Content Key Policy for AES-128 encryption
            var contentKeyPolicy = await CreateContentKeyPolicyAsync(mediaService, resourceGroupName, accountName, contentKeyPolicyName, issuer, audience);
            // Create Streaming Locator with Content Key Policy
            var aesLocator = await CreateAES128StreamingLocatorAsync(mediaService, resourceGroupName, accountName, outputAsset.Name, aes128locatorName, contentKeyPolicy.Name);
            // List url for encrypted streaming
            var aesPaths = await mediaService.StreamingLocators.ListPathsAsync(resourceGroupName, accountName, aes128locatorName);

            Console.WriteLine();
            Console.WriteLine("The following URL is available for adaptive streaming with AES-128 encryption:");
            foreach (StreamingPath path in aesPaths.StreamingPaths)
            {
                foreach (string streamingFormatPath in path.Paths)
                {
                    var streamingUrl = $"https://{streamingEndpoint.HostName}{streamingFormatPath}";
                    Console.WriteLine($"{path.StreamingProtocol}: {streamingUrl}");
                    streamingUrls.Add(streamingUrl);
                }
            }

            // Generate test token for AES-128 encryption
            ContentKeyPolicyTokenRestriction restriction = (ContentKeyPolicyTokenRestriction)contentKeyPolicy.Options.First().Restriction;
            ContentKeyPolicySymmetricTokenKey tokenKey = (ContentKeyPolicySymmetricTokenKey)restriction.PrimaryVerificationKey;
            var token = JWT.Encode(new Dictionary<string, object>
                {
                    { "exp", DateTimeOffset.UtcNow.AddHours(6).ToUnixTimeSeconds() },
                    { "nbf", DateTimeOffset.UtcNow.ToUnixTimeSeconds() },
                    { "iss", issuer },
                    { "aud", audience }
                },
                tokenKey.KeyValue,
                JwsAlgorithm.HS256);
            Console.WriteLine($"Token for encrypted playback (valid for 6hrs): {token}");
        }

        private async Task<AzureMediaServicesClient> CreateAmsClient()
        {
            ClientCredential clientCredential = new ClientCredential(_azureOptions.ClientId, _azureOptions.ClientSecret);
            ServiceClientCredentials serviceClientCredentials =
                await ApplicationTokenProvider.LoginSilentAsync(
                    _azureOptions.AadTenantId,
                    clientCredential,
                    ActiveDirectoryServiceSettings.Azure);
            var client = new HttpClient();
            client.BaseAddress = new Uri(_azureOptions.ApiEndpoint);

            return new AzureMediaServicesClient(serviceClientCredentials, client, true)
            {
                SubscriptionId = _azureOptions!.SubscriptionId
            };
        }

        private AzureMediaServicesClient CreateRmsClient()
        {
            var serviceClientCredentials = new RmsApiKeyCredentials(
                new Uri(_rmsOptions.ApiEndpoint),
                _rmsOptions.SubscriptionId,
                _rmsOptions.ApiKey);

            return new AzureMediaServicesClient(serviceClientCredentials, new HttpClient(), true)
            {
                SubscriptionId = _rmsOptions!.SubscriptionId
            };
        }

        private static Task<StreamingLocator> CreateStreamingLocatorAsync(
            AzureMediaServicesClient mediaService,
            string resourceGroupName,
            string accountName,
            string assetName,
            string locatorName)
        {
            return mediaService.StreamingLocators.CreateAsync(
                resourceGroupName,
                accountName,
                locatorName,
                new StreamingLocator
                {
                    AssetName = assetName,
                    StreamingPolicyName = "Predefined_DownloadAndClearStreaming",
                });
        }

        private static Task<StreamingLocator> CreateAES128StreamingLocatorAsync(
            AzureMediaServicesClient mediaService,
            string resourceGroupName,
            string accountName,
            string assetName,
            string locatorName,
            string contentKeyPolicyName)
        {
            return mediaService.StreamingLocators.CreateAsync(
                resourceGroupName,
                accountName,
                locatorName,
                new StreamingLocator
                {
                    AssetName = assetName,
                    StreamingPolicyName = "Predefined_ClearKey",
                    DefaultContentKeyPolicyName = contentKeyPolicyName,
                });
        }

        private static async Task<Job> WaitForJobToFinishAsync(AzureMediaServicesClient mediaService, string resourceGroup, string accountName, string transformName, string jobName)
        {
            var sleepInterval = TimeSpan.FromSeconds(10);
            JobState? state;
            int progress = 0;

            Job job;
            do
            {
                job = await mediaService.Jobs.GetAsync(resourceGroup, accountName, transformName, jobName);
                state = job?.State;
                progress = job?.Outputs.FirstOrDefault()?.Progress ?? 0;

                Console.WriteLine($"Job state: {state}, progress: {progress}");
                if (state != JobState.Finished && state != JobState.Error && state != JobState.Canceled)
                {
                    await Task.Delay(sleepInterval);
                }
            }
            while (state != JobState.Finished && state != JobState.Error && state != JobState.Canceled);

            return job;
        }

        private static async Task<Job> SubmitJobAsync(
            AzureMediaServicesClient mediaService,
            string resourceGroupName,
            string accountName,
            string transformName,
            string jobName,
            JobInput input,
            string outputAssetName)
        {
            var jobParameters = new Job
            {
                Input = input,
                Outputs = new List<JobOutput>()
                {
                    new JobOutputAsset(outputAssetName),
                },
            };

            var job = await mediaService.Jobs.CreateAsync(
                resourceGroupName,
                accountName,
                transformName,
                jobName,
                jobParameters);

            return job;
        }

        private static async Task UploadFileToAsset(
            AzureMediaServicesClient mediaService,
            string resourceGroupName,
            string accountName,
            string assetName,
            string fileName)
        {
            var sasUriCollection = await mediaService.Assets.ListContainerSasAsync(
                resourceGroupName,
                accountName,
                assetName,
                AssetContainerPermission.ReadWrite,
                DateTime.UtcNow.AddHours(1));

            var container = new BlobContainerClient(new Uri(sasUriCollection.AssetContainerSasUrls.First()));
            BlobClient blob = container.GetBlobClient(Path.GetFileName(fileName));

            await blob.UploadAsync(fileName);
        }

        private static async Task<Asset> CreateAsset(
            AzureMediaServicesClient mediaService,
            string resourceGroupName,
            string accountName,
            string assetName)
        {
            var asset = await mediaService.Assets.CreateOrUpdateAsync(
                resourceGroupName,
                accountName,
                assetName,
                new Asset());

            return asset;
        }

        private async Task<ContentKeyPolicy> CreateContentKeyPolicyAsync(
            AzureMediaServicesClient mediaService,
            string resourceGroupName,
            string accountName,
            string policyName,
            string issuer,
            string audience)
        {
            var options = new ContentKeyPolicyOption[]
            {
                new ContentKeyPolicyOption(
                    name: "ae128-option",
                    configuration: new ContentKeyPolicyClearKeyConfiguration(),
                    restriction: new ContentKeyPolicyTokenRestriction(
                        issuer: issuer,
                        audience: audience,
                        primaryVerificationKey: new ContentKeyPolicySymmetricTokenKey(Guid.NewGuid().ToByteArray()),
                        restrictionTokenType: ContentKeyPolicyRestrictionTokenType.Jwt)
                )
            };

            return await mediaService.ContentKeyPolicies.CreateOrUpdateAsync(resourceGroupName, accountName, policyName, options, "test policy description");
        }

        private static async Task<Transform> CreateTransformAsync(
            AzureMediaServicesClient mediaService,
            string resourceGroupName,
            string accountName,
            string transformName)
        {
            var outputs = new List<TransformOutput>
            {
                new TransformOutput
                {
                    Preset = new StandardEncoderPreset
                    {
                        Codecs = new List<Codec>
                        {
                            new AacAudio
                            {
                                Channels = 2,
                                SamplingRate = 48000,
                                Bitrate = 128000,
                                Profile = AacAudioProfile.AacLc,
                            },
                            new H264Video()
                            {
                                KeyFrameInterval = TimeSpan.FromSeconds(2),
                                Layers = new List<H264Layer>
                                {
                                    new H264Layer(bitrate: 3600000)
                                    {
                                        Width = "1280",
                                        Height = "720",
                                        Label = "HD-3600kbps",
                                    },
                                    new H264Layer(bitrate: 1600000)
                                    {
                                        Width = "960",
                                        Height = "540",
                                        Label = "SD-1600kbps",
                                    },
                                    new H264Layer(bitrate: 600000)
                                    {
                                        Width = "640",
                                        Height = "360",
                                        Label = "SD-600kbps",
                                    },
                                },
                            },
                            new JpgImage(start: "25%")
                            {
                                Start = "25%",
                                Step = "25%",
                                Range = "80%",
                                Layers = new List<JpgLayer>
                                {
                                    new JpgLayer
                                    {
                                        Width = "50%",
                                        Height = "50%",
                                    },
                                },
                            },
                        },
                        Formats = new List<Format>
                        {
                            new Mp4Format(filenamePattern: "Video-{Basename}-{Label}-{Bitrate}{Extension}"),
                            new JpgFormat(filenamePattern: "Thumbnail-{Basename}-{Index}{Extension}"),
                        },
                        Filters = new Filters
                        {
                            Overlays = new List<Overlay>
                            {
                                new VideoOverlay(LogoMaskLabel)
                                {
                                    Start = TimeSpan.FromSeconds(3),
                                    End = TimeSpan.FromSeconds(10),
                                    Position = new Rectangle (left:"75%", top:"75%", width:"25%", height: "25%"),
                                    FadeInDuration = TimeSpan.FromSeconds(2),
                                }
                            }
                        }
                    },
                }
            };

            var transform = await mediaService.Transforms.CreateOrUpdateAsync(
                resourceGroupName,
                accountName,
                transformName,
                outputs,
                "A simple custom encoding transforms.");

            return transform;
        }

        private static bool IsUrl(string path) => Uri.TryCreate(path, UriKind.Absolute, out Uri? uri) && (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps);
    }
}

File: node-js-demo/VodProvider.js
const { AzureMediaServices: MediaServices } = require("@azure/arm-mediaservices");
const { BlobServiceClient } = require("@azure/storage-blob");
const { setLogLevel } = require("@azure/logger");
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const fs = require('fs');
const fetch = require('node-fetch');
const RmsApiKeyCredentials = require('./RmsApiKeyCredentials');

// Uncomment the line below to enable verbose logging if needed
// setLogLevel("verbose");

class VodProvider {
  transformName = 'RmsTestTransform';
  streamingEndpointName = 'default';
  options = {};

  constructor(options) {
    this.options = options;
  }

  async createRMSClient() {
    const credentials = new RmsApiKeyCredentials(
      this.options.ApiEndpoint,
      this.options.SubscriptionId,
      this.options.ApiKey
    );

    const mediaService = new MediaServices(credentials, this.options.SubscriptionId, {
      endpoint: this.options.ApiEndpoint
    });

    return mediaService;
  }

  async createAssetsAndUploadFile(mediaService, inputFile, inputAssetName, outputAssetName) {
    console.log('Creating input and output assets...');
    const inputAsset = await mediaService.assets.createOrUpdate(this.options.ResourceGroupName, this.options.MediaServicesAccountName, inputAssetName, {});
    console.log('Input asset created: ', inputAsset.name);
    const outputAsset = await mediaService.assets.createOrUpdate(this.options.ResourceGroupName, this.options.MediaServicesAccountName, outputAssetName, {});
    console.log('Output asset created: ', outputAsset.name);

    console.log('Uploading file...');
    const asset = await mediaService.assets.get(this.options.ResourceGroupName, this.options.MediaServicesAccountName, inputAssetName);
    const listContainerSas = await mediaService.assets.listContainerSas(this.options.ResourceGroupName, this.options.MediaServicesAccountName, asset.name, { permissions: 'ReadWrite' });
    const container = listContainerSas.assetContainerSasUrls[0];
    const blobService = new BlobServiceClient(container);
    const containerClient = blobService.getContainerClient(asset.name);
    const blobClient = containerClient.getBlobClient(inputFile);
    const blockBlobClient = blobClient.getBlockBlobClient();

    await blockBlobClient.uploadFile(inputFile);

    return { inputAsset, outputAsset };
  }

  async createJob(mediaService, jobName, inputAsset, outputAsset) {
    console.log('Creating job...');

    const jobInput = {
      odataType: '#Microsoft.Media.JobInputAsset',
      assetName: inputAsset.name
    };

    const jobOutput = {
      odataType: '#Microsoft.Media.JobOutputAsset',
      assetName: outputAsset.name
    };

    const job = await mediaService.jobs.create(this.options.ResourceGroupName, this.options.MediaServicesAccountName, this.transformName, jobName, {
      input: jobInput,
      outputs: [jobOutput],
    });

    console.log(`Job created: ${job.name}`);

    let jobStatus = await mediaService.jobs.get(this.options.ResourceGroupName, this.options.MediaServicesAccountName, this.transformName, jobName);

    while (jobStatus.state !== 'Finished' && jobStatus.state !== 'Error') {
      if (jobStatus.state === 'Scheduled' || jobStatus.state === 'Processing') {
        console.log(`Job status: ${jobStatus.state}. Progress: ${jobStatus.outputs[0].progress}%`);
      }

      await new Promise(resolve => setTimeout(resolve, 10000));
      jobStatus = await mediaService.jobs.get(this.options.ResourceGroupName, this.options.MediaServicesAccountName, this.transformName, jobName);
    }

    if (jobStatus.state === 'Error') {
      console.error(`Job ${jobStatus.name} failed:`, jobStatus.error.message);
    }

    console.log(`Job ${jobStatus.name} finished successfully.`);
  }

  async createStreamingLocator(mediaService, outputAsset, locatorName) {
    // create streaming locator
    console.log('Creating streaming locator...');
    const locator = await mediaService.streamingLocators.create(this.options.ResourceGroupName, this.options.MediaServicesAccountName, locatorName, {
      assetName: outputAsset.name,
      streamingPolicyName: 'Predefined_ClearStreamingOnly'
    });

    console.log(`Streaming locator created: ${locator.name}`)

    // get streaming endpoint
    const streamingEndpoint = await mediaService.streamingEndpoints.get(this.options.ResourceGroupName, this.options.MediaServicesAccountName, this.streamingEndpointName);

    if (streamingEndpoint.resourceState !== 'Running') {
      console.log('Streaming endpoint is not running, starting it...');
      await mediaService.streamingEndpoints.start(this.options.ResourceGroupName, this.options.MediaServicesAccountName, this.streamingEndpointName);
      console.log('Streaming endpoint started successfully.');
    }

    // List streaming locator for asset
    const streamingLocators = await mediaService.streamingLocators.listPaths(this.options.ResourceGroupName, this.options.MediaServicesAccountName, locatorName);

    // List url for streaming locator
    const paths = streamingLocators.streamingPaths.map(path => path.paths[0]);
    const downloadPaths = streamingLocators.downloadPaths;
    const downloadUrls = [];
    const streamingUrls = [];

    for (const path of paths) {
      const url = `${streamingEndpoint.hostName}/${path}`;
      streamingUrls.push(url);
    }

    for (const path of downloadPaths) {
      const url = `${streamingEndpoint.hostName}/${path}`;
      downloadUrls.push(url);
    }

    console.log('The following URLs are available for adaptive streaming:');
    streamingUrls.forEach(url => console.log(url));

    console.log('The following URLs are available for download:');
    downloadUrls.forEach(url => console.log(url));

    return streamingEndpoint;
  }

  async createAES128StreamingLocator(mediaService, outputAsset, locatorName, contentKeyPolicyName, streamingEndpoint) {
    console.log('Creating Content Key Policy for AES-128 encryption...');

    const issuer = 'ravnur';
    const audience = 'rmstest';
    const password = 'banana strawberry apple pear' // Change this to a strong password

    const configuration = {
      odataType: "#Microsoft.Media.ContentKeyPolicyClearKeyConfiguration",
    }

    const primaryVerificationKey = {
      odataType: "#Microsoft.Media.ContentKeyPolicySymmetricTokenKey",
      keyValue: this.deriveKey(password),
    }

    const restriction = {
      odataType: '#Microsoft.Media.ContentKeyPolicyTokenRestriction',
      issuer,
      audience,
      restrictionTokenType: 'Jwt',
      primaryVerificationKey,
    };

    const contentKeyPolicy = await mediaService.contentKeyPolicies.createOrUpdate(this.options.ResourceGroupName, this.options.MediaServicesAccountName, contentKeyPolicyName, {
      options: [{
        configuration,
        restriction
      }],
    });

    console.log(`Content Key Policy created: ${contentKeyPolicy.name}`)

    // Create Streaming Locator with Content Key Policy
    await mediaService.streamingLocators.create(this.options.ResourceGroupName, this.options.MediaServicesAccountName, locatorName, {
      assetName: outputAsset.name,
      streamingPolicyName: 'Predefined_ClearKey',
      defaultContentKeyPolicyName: contentKeyPolicy.name,
    });

    console.log(`AES-128 Streaming Locator created: ${locatorName}`);

    const aes128StreamingLocators = await mediaService.streamingLocators.listPaths(this.options.ResourceGroupName, this.options.MediaServicesAccountName, locatorName);
    const aes128Paths = aes128StreamingLocators.streamingPaths.map(path => path.paths[0]);
    const aes128StreamingUrls = [];

    for (const path of aes128Paths) {
      const url = `${streamingEndpoint.hostName}/${path}`;
      aes128StreamingUrls.push(url);
    }

    console.log('The following URLs are available for adaptive streaming with AES-128 encryption:');
    aes128StreamingUrls.forEach(url => console.log(url));

    // // Define token payload
    const tokenPayload = {
      exp: Math.floor(Date.now() / 1000) + (6 * 60 * 60), // 6 hours expiration
      nbf: Math.floor(Date.now() / 1000),
      iss: issuer,
      aud: audience,
    };

    const tokenKey = this.deriveKey(password)

    // Sign the token
    const token = jwt.sign(tokenPayload, tokenKey, { algorithm: 'HS256' });
    console.log(`Token for encrypted playback (valid for 6hrs): ${token}`);
  }

  async downloadFileIfNeeded(inputFile) {
    let filePath = inputFile;

    if (this.isUrl(filePath)) {
      console.log('Input file is a URL, downloading...');
      const url = new URL(filePath);
      const fileName = `${new Date().getTime()}-${url.pathname.split('/').pop()}`
      const path = `./input/${fileName}`;
      
      try {
        const response = await fetch(filePath);
        if (!response.ok) {
          throw new Error(`Failed to download file: ${response.statusText}`);
        }
        const buffer = await response.buffer();
        fs.writeFileSync(path, buffer);
        console.log(`File downloaded and saved to ${path}`);

        // Fix the input file path
        filePath = path;
      } catch (error) {
        console.error('Error downloading file:', error);
      }
    }

    return filePath;
  }

  async createVod(inputFile) {
    console.log(`Creating VOD for ${inputFile} using RMS`);

    const unique = new Date().getTime();
    const inputAssetName = `input-${unique}`;
    const outputAssetName = `output-${unique}`;
    const jobName = `job-${unique}`;
    const locatorName = `locator-${unique}`;
    const aes128locatorName = `aes128locator-${unique}`;
    const contentKeyPolicyName = `contentKeyPolicy-${unique}`;
    const filePath = await this.downloadFileIfNeeded(inputFile);

    try {
      // Create RMS client
      const mediaService = await this.createRMSClient();

      // Create transform if it doesn't exist
      await this.createTransform(mediaService, this.transformName);

      // Create input asset and upload file
      const { inputAsset, outputAsset } = await this.createAssetsAndUploadFile(mediaService, filePath, inputAssetName, outputAssetName);

      // Create the job and wait for it to finish
      await this.createJob(mediaService, jobName, inputAsset, outputAsset);

      // Create Streaming Locator
      const streamingEndpoint = await this.createStreamingLocator(mediaService, outputAsset, locatorName);

      // Create AES128 Streaming Locator
      await this.createAES128StreamingLocator(mediaService, outputAsset, aes128locatorName, contentKeyPolicyName, streamingEndpoint);
    } catch (err) {
      console.error('Error creating VOD:', err.message);
    } finally {
      process.exit();
    }
  }

  deriveKey(password) {
    let hash = crypto.createHash('sha256');
    hash.update(new TextEncoder().encode(password));
    return hash.digest().subarray(0, 16)
  }

  async createTransform(client) {
    console.log('Checking if Transform exists...');
    let transform = null;

    try {
      transform = await client.transforms.get(this.options.ResourceGroupName, this.options.MediaServicesAccountName, this.transformName);
      console.log('Transform already exists.');
    } catch (error) {
      if (error.statusCode === 404) {
        console.log('Transform not found, creating...');
        const TRANSFORM_DATA = require('./transform-data.json');
        transform = await client.transforms.createOrUpdate(this.options.ResourceGroupName, this.options.MediaServicesAccountName, this.transformName, TRANSFORM_DATA);
        console.log('Transform created successfully.');
      } else {
        throw error; // Rethrow the error if it's not a 404
      }
    }
    return transform;
  }

  async isUrl(url) {
    return url.includes('http://') || url.includes('https://');
  }
}

module.exports = VodProvider;